File: src/queue/queue.ts
import { redisClient } from './redis';
import { config } from '../config';
import { TaskPriority, QueueMessage } from '../types';

// Maps priority → Redis list key. LPUSH enqueues, BRPOP dequeues from right.
// Workers poll high → normal → low in order, giving strict priority.
function queueKey(priority: TaskPriority): string {
  switch (priority) {
    case 'high': return config.queue.highPriority;
    case 'low': return config.queue.lowPriority;
    default: return config.queue.normalPriority;
  }
}

export async function enqueue(taskId: string, type: string, priority: TaskPriority): Promise<void> {
  const message: QueueMessage = {
    task_id: taskId,
    type: type as QueueMessage['type'],
    priority,
    enqueued_at: new Date().toISOString(),
  };
  const key = queueKey(priority);
  // LPUSH pushes to left; BRPOP pops from right → FIFO within same priority
  await redisClient.lPush(key, JSON.stringify(message));
}

export async function dequeue(timeoutSeconds: number): Promise<QueueMessage | null> {
  // Poll high, normal, low in strict priority order (BRPOP is atomic and blocking)
  const result = await redisClient.brPop(
    [config.queue.highPriority, config.queue.normalPriority, config.queue.lowPriority],
    timeoutSeconds
  );
  if (!result) return null;
  return JSON.parse(result.element) as QueueMessage;
}

export async function enqueueDeadLetter(taskId: string, reason: string): Promise<void> {
  const message = { task_id: taskId, reason, dead_at: new Date().toISOString() };
  await redisClient.lPush(config.queue.deadLetter, JSON.stringify(message));
}

export async function getQueueDepths(): Promise<Record<string, number>> {
  const [high, normal, low, dead] = await Promise.all([
    redisClient.lLen(config.queue.highPriority),
    redisClient.lLen(config.queue.normalPriority),
    redisClient.lLen(config.queue.lowPriority),
    redisClient.lLen(config.queue.deadLetter),
  ]);
  return { high, normal, low, dead, total: high + normal + low };
}

File: app/queue/queue.py
import json
from typing import Optional
from app.core.config import get_settings
from app.queue.client import get_async_redis, get_sync_redis

settings = get_settings()

PRIORITY_KEYS = {
    "high":   settings.QUEUE_HIGH,
    "normal": settings.QUEUE_NORMAL,
    "low":    settings.QUEUE_LOW,
}


def _queue_key(priority: str) -> str:
    return PRIORITY_KEYS.get(priority, settings.QUEUE_NORMAL)


# ── Async (used by API) ───────────────────────────────────────────────────────

async def enqueue(task_id: str, task_type: str, priority: str) -> None:
    r = get_async_redis()
    message = json.dumps({"task_id": task_id, "type": task_type, "priority": priority})
    await r.lpush(_queue_key(priority), message)


async def get_queue_depths() -> dict:
    r = get_async_redis()
    high, normal, low, dead = await r.llen(settings.QUEUE_HIGH), \
        await r.llen(settings.QUEUE_NORMAL), \
        await r.llen(settings.QUEUE_LOW), \
        await r.llen(settings.QUEUE_DEAD)
    return {"high": high, "normal": normal, "low": low, "dead": dead, "total": high + normal + low}


# ── Sync (used by worker threads) ────────────────────────────────────────────

def dequeue_sync(timeout: int) -> Optional[dict]:
    """BRPOP across priority queues — returns parsed message or None on timeout."""
    r = get_sync_redis()
    result = r.brpop(
        [settings.QUEUE_HIGH, settings.QUEUE_NORMAL, settings.QUEUE_LOW],
        timeout=timeout,
    )
    if not result:
        return None
    _, raw = result
    return json.loads(raw)


def enqueue_dead_letter_sync(task_id: str, reason: str) -> None:
    r = get_sync_redis()
    r.lpush(settings.QUEUE_DEAD, json.dumps({"task_id": task_id, "reason": reason}))


def enqueue_sync(task_id: str, task_type: str, priority: str) -> None:
    r = get_sync_redis()
    message = json.dumps({"task_id": task_id, "type": task_type, "priority": priority})
    r.lpush(_queue_key(priority), message)
